src/dtx/dtx_common.c

```diff
@@ -663,3 +656,211 @@ dtx_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
+	return rc;
+}
+
+void
+dtx_exec_arg_free(struct dtx_exec_arg *exec_arg)
+{
+	ABT_future_free(&exec_arg->future);
+	D_FREE(exec_arg);
+}
+
+static int
+dtx_wait(struct dtx_handle *dth, struct dtx_conflict_entry **dces)
+{
+	struct dtx_exec_arg	*exec_arg = dth->dth_exec_arg;
+	int			rc;
+
+	if (exec_arg == NULL)
+		return 0;
+
+	D_ASSERT(exec_arg->future != ABT_FUTURE_NULL);
+	rc = ABT_future_wait(exec_arg->future);
+	D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_wait failed %d.\n", rc);
+	rc = exec_arg->exec_result;
+	if (rc == -DER_INPROGRESS && dces != NULL) {
+		struct dtx_exec_shard_arg	*shard_arg;
+		struct dtx_conflict_entry	*conflict;
+		int				shard_cnt = exec_arg->shard_cnt;
+		int				i;
+		int				j;
+
+		D_ALLOC_ARRAY(conflict, shard_cnt);
+		if (conflict == NULL) {
+			rc = -DER_NOMEM;
+			goto out;
+		}
+
+		shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
+		for (i = 0, j = 0; i < shard_cnt; i++, shard_arg++) {
+			if (!daos_is_zero_dti(&shard_arg->exec_dce.dce_xid)) {
+				daos_dti_copy(&conflict[j].dce_xid,
+					      &shard_arg->exec_dce.dce_xid);
+				conflict[j++].dce_dkey =
+						shard_arg->exec_dce.dce_dkey;
+			}
+		}
+
+		D_ASSERT(j > 0);
+
+		*dces = conflict;
+		rc = j;
+	}
+
+out:
+	D_DEBUG(DB_TRACE, "dth "DF_DTI" rc %d\n", DP_DTI(&dth->dth_xid), rc);
+	dtx_exec_arg_free(exec_arg);
+
+	return rc;
+};
+
+void
+dtx_exec_ops_ult(void *arg)
+{
+	struct dtx_exec_arg	*exec_arg = arg;
+	ABT_future		future = exec_arg->future;
+	struct dtx_exec_shard_arg *shard_arg;
+	uint32_t		i;
+	int			shard_cnt = exec_arg->shard_cnt;
+	int			rc = 0;
+
+	D_ASSERT(shard_cnt >= 1);
+	D_ASSERT(future != ABT_FUTURE_NULL);
+	shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
+	for (i = 0; i < shard_cnt; i++, shard_arg++) {
+		rc = dtx_exec_op(shard_arg, i);
+		if (rc != 0)
+			break;
+	}
+
+	if (rc != 0) {
+		D_ASSERT(i < shard_cnt);
+		for (i++; i < shard_cnt; i++) {
+			shard_arg++;
+			shard_arg->exec_shard_rc = rc;
+			rc = ABT_future_set(future, shard_arg);
+			D_ASSERTF(rc == ABT_SUCCESS,
+				  "ABT_future_set failed %d.\n", rc);
+		}
+	}
+}
+
+/**
+ * Execute the operations on all targets.
+ */
+int
+dtx_exec_ops(struct daos_shard_tgt *shard_tgts, int tgts_cnt,
+	     struct dtx_handle *dth, dtx_exec_shard_func_t exec_func,
+	     void *func_arg)
+{
+	struct dtx_exec_arg		*exec_arg;
+	struct dtx_exec_shard_arg	*shard_arg;
+	ABT_future			future;
+	int				i;
+	int				rc;
+
+	D_ALLOC(exec_arg, sizeof(struct dtx_exec_arg) +
+		tgts_cnt * sizeof(struct dtx_exec_shard_arg));
+	if (exec_arg == NULL)
+		return -DER_NOMEM;
+
+	rc = ABT_future_create(tgts_cnt, dtx_exec_ops_comp_cb, &future);
+	if (rc != ABT_SUCCESS) {
+		D_ERROR("ABT_future_create failed %d.\n", rc);
+		D_FREE(exec_arg);
+		return dss_abterr2der(rc);
+	}
+
+	exec_arg->exec_func	= exec_func;
+	exec_arg->exec_func_arg	= func_arg;
+	exec_arg->future	= future;
+	exec_arg->shard_cnt	= tgts_cnt;
+	exec_arg->dth		= dth;
+	shard_arg = (struct dtx_exec_shard_arg *)(exec_arg + 1);
```

Comment:
use a struct rather than + 1 for less obfuscation

Reply:
ok